git-worktree/src/lib.rs

```diff
@@ -1 +1,62 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use anyhow::Result;
+use git_hash::oid;
+use git_index::{entry::Mode, State};
+use git_object::Data;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::str;
+
+mod options;
+
+pub use options::Options;
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &State, path: P, mut find: Find, opts: Options) -> Result<()>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<Data<'a>>,
+{
+    let path = path.as_ref();
+    let entries = state.entries();
+    let mut buf = Vec::new();
+    for entry in entries {
+        let dest = path.join(entry.path(state).to_string());
+        create_dir_all(dest.parent().expect("Path has no parent"))?;
+        match entry.mode {
+            Mode::FILE | Mode::FILE_EXECUTABLE => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                std::fs::write(dest, obj.data)?;
+            }
+            Mode::SYMLINK => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                let linked_to = str::from_utf8(obj.data)?;
+                if opts.symlinks {
+                    #[cfg(unix)]
+                    std::os::unix::fs::symlink(linked_to, dest)?;
+                    #[cfg(windows)]
+                    if dest.exists() {
+                        if dest.is_file() {
+                            std::os::windows::fs::symlink_file(linked_to, dest)?;
+                        } else {
+                            std::os::windows::fs::symlink_dir(linked_to, dest)?;
+                        }
+                    }
+                } else {
+                    let linked_to_path = path.join(linked_to);
+                    if linked_to_path.exists() {
+                        std::fs::copy(linked_to_path, dest)?;
+                    } else {
+                        std::fs::write(dest, linked_to)?;
+                    }
+                }
+            }
+            Mode::DIR => todo!(),
+            Mode::COMMIT => todo!(),
+            _ => {}
```

Comment:
Probably it's best to put a `todo()` or `unreachable!()` here as all cases should be clear in code (eventually).